### PR TITLE
Re-write trainer as a bunch of callback hooks

### PR DESCRIPTION
**Pull request by relh**
Trainer should just train. Everything else should not be mentioned in the trainer code.

Trainer(config: TrainerConfig, curriculum: Curriculum, policy: Policy)

Everything else should happen externally. (wandb, checkpointing, evals, stats, heartbeats [ideally perf])

### CODE REVIEW

File: metta/rl/hooks/base.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    __slots__ = (
+        "trainer_cfg",
+        "device",
+        "instance_name",
+        "critical",
+    )
+
+    def __init__(
+        self,
+        trainer_cfg,
```
Comment (daveey) [inline metta/rl/hooks/base.py:25]: : TrainerConfig ?

File: metta/rl/loss/base_loss.py [hunk 2]
```diff
@@ -119,6 +119,14 @@ def on_mb_end(self, trainer_state: TrainerState) -> None:
     def on_train_phase_end(self, trainer_state: TrainerState) -> None:
         """We've completed the train phase and will be transitioning to the next rollout phase."""
 
+    def on_rollout_end(self, trainer_state: TrainerState) -> None:
```
Comment (daveey) [inline metta/rl/loss/base_loss.py:122]: Should this just subclass Hook and not overload most noop methods?

File: metta/rl/trainer.py [hunk 3]
```diff
@@ -209,6 +209,9 @@ def train(
     # Instantiate configured composable losses dynamically
     loss_instances = trainer_cfg.losses.init_losses(policy, trainer_cfg, vecenv, device, checkpoint_manager)
```
Comment (daveey) [inline metta/rl/trainer.py:210]: _losses: list[BaseLoss]

The losses should be passed in or we should have a self._create_losses()

**Comment by relh** [comment]
This is still a draft, let me make it one

File: metta/rl/trainer.py (hunk 3 above)
Comment (daveey) [inline metta/rl/trainer.py:210]: losses: List[Loss]

**Review by daveey — APPROVED** [review]
This would be easier to review if you included at least one hook. 

I'm still not convinced we want TrainerState, train() instead of Trainer.train()
